File: c_src/interpreter.c
#include <stdio.h>

#include "gc.h"
#include "object.h"
#include "builtins.h"
#include "errors.h"
#include "thread.h"

typedef enum Opcode {
	ERROR = 0,
	ST_SWAP = 1,
	ST_POP = 2,
	ST_DUP = 3,
	ST_DUP2 = 4,
	LIT_BYTES = 10,
	LIT_INT = 11,
	LIT_FLOAT = 12,
	LIT_SLICE = 13,
	LIT_NONE = 14,
	LIT_TRUE = 15,
	LIT_FALSE = 16,
	TUPLE_0 = 17,
	TUPLE_1 = 18,
	TUPLE_2 = 19,
	TUPLE_3 = 20,
	TUPLE_4 = 21,
	TUPLE_N = 22,
	CLOSURE = 23,
	CLOSURE_BIND = 24,
	EMPTY_DICT = 25,
	CLASS = 26,
	GET_ATTR = 40,
	SET_ATTR = 41,
	DEL_ATTR = 42,
	GET_ITEM = 43,
	SET_ITEM = 44,
	DEL_ITEM = 45,
	GET_LOCAL = 46,
	SET_LOCAL = 47,
	DEL_LOCAL = 48,
	LOAD_ARGS = 49,
	JUMP = 60,
	JUMP_IF = 61,
	TRY = 62,
	TRY_END = 63,
	CALL = 64,
	SPAWN = 65,
	RAISE = 66,
	RETURN = 67,
	YIELD = 68,
	RAISE_IF_NOT_STOP = 69, // I swear to god I didn't give the funniest opcode the funniest number on purpose
	OP_ADD = 80,
	OP_SUB = 81,
	OP_MUL = 82,
	OP_DIV = 83,
	OP_MOD = 84,
	OP_AND = 85,
	OP_OR = 86,
	OP_XOR = 87,
	OP_NEG = 88,
	OP_NOT = 89,
	OP_INV = 90,
	OP_EQ = 91,
	OP_NE = 92,
	OP_GT = 93,
	OP_LT = 94,
	OP_GE = 95,
	OP_LE = 96,
	OP_SHL = 97,
	OP_SHR = 98,
} Opcode;

#define ENSURE_BYTES(n) ((size_t)(*pointer - bytes_data(bytecode) + (n)) <= bytecode->len)

Opcode next_opcode(BytesObject *bytecode, const char **pointer) {
	if (!ENSURE_BYTES(1)) {
		return ERROR;
	}
	Opcode result = **pointer;
	(*pointer)++;
	return result;
}

bool next_num_unsigned(BytesObject *bytecode, const char **pointer, uint64_t *out) {
	// https://en.wikipedia.org/wiki/LEB128
	uint64_t result = 0;
	uint64_t shift = 0;
	while (true) {
		if (!ENSURE_BYTES(1)) {
			return false;
		}
		unsigned char next = **pointer;
		(*pointer)++;
		result |= (uint64_t)(next & 0x7f) << shift;
		shift += 7;
		if ((next & 0x80) == 0) {
			*out = result;
			return true;
		}
	}
}

bool next_num_signed(BytesObject *bytecode, const char **pointer, int64_t *out) {
	int64_t result = 0;
	uint64_t shift = 0;
	while (true) {
		if (!ENSURE_BYTES(1)) {
			return false;
		}
		unsigned char next = **pointer;
		(*pointer)++;
		result |= (uint64_t)(next & 0x7f) << shift;
		shift += 7;
		if ((next & 0x80) == 0) {
			if (shift < 64 && (next & 0x40)) {
				result |= ~0 << shift;
			}
			*out = result;
			return true;
		}
	}
}

bool next_float(BytesObject *bytecode, const char **pointer, double *out) {
	if (!ENSURE_BYTES(sizeof(double))) {
		return false;
	}
	*out = *(double*)*pointer;
	*pointer += sizeof(double);
	return true;
}

bool next_offset(BytesObject *bytecode, const char **pointer, size_t *out) {
	if (!ENSURE_BYTES(sizeof(uint32_t))) {
		return false;
	}
	*out = *(uint32_t*)*pointer;
	*pointer += sizeof(uint32_t);
	return true;
}

BytesUnownedObject *next_bytes(BytesObject *bytecode, const char **pointer) {
	uint64_t _size;
	if (!next_num_unsigned(bytecode, pointer, &_size)) {
		return NULL;
	}
	size_t size = _size;
	if (!ENSURE_BYTES(size)) {
		return NULL;
	}
	BytesUnownedObject *result = bytes_unowned_raw(*pointer, size, (Object*)bytecode);
	*pointer += size;
	return result;
}

Object *interpreter(ClosureObject *closure, TupleObject *args) {
	DictObject *locals = dict_dup_inner(closure->context);
	if (locals == NULL) {
		return NULL;
	}
	gc_root((Object*)locals);

	ListObject *stack = list_raw(NULL, 0);
	if (stack == NULL) {
		gc_unroot((Object*)locals);
		return NULL;
	}
	gc_root((Object*)stack);

	ListObject *temproot = list_raw(NULL, 0);
	if (temproot == NULL) {
		gc_unroot((Object*)stack);
		gc_unroot((Object*)locals);
		return NULL;
	}
	gc_root((Object*)temproot);

	ListObject *trystack = list_raw(NULL, 0);
	if (trystack == NULL) {
		gc_unroot((Object*)stack);
		gc_unroot((Object*)locals);
		gc_unroot((Object*)temproot);
		return NULL;
	}
	gc_root((Object*)trystack);

	const char *pointer = bytes_data(closure->bytecode);
	Object *result = NULL;

	while (true) {
		while (temproot->len) {
			if (!list_pop_back_inner(temproot)) {
				puts("Fatal error: could not pop temproot");
				exit(1);
			}
		}
		gc_probe();
		if (!gil_probe()) {
			goto ERROR;
		}
		if (pointer == &bytes_data(closure->bytecode)[closure->bytecode->len]) {
			result = (Object*)&g_none;
			break;
		}
		Opcode opcode = next_opcode(closure->bytecode, &pointer);
		switch (opcode) {

#define CHECK(_val) ({ __typeof__(_val) _evaluated = (_val); if (!_evaluated) { break; } _evaluated; })
#define POP() ({ Object *_popped = list_pop_back_inner(stack); if (_popped == NULL) { puts("Fatal error: stack underflow"); exit(1); } _popped; })
#define PUSH(_pushed) ({ if (!list_push_back_inner(stack, (Object*)_pushed)) { break; } })
#define TEMPROOT(_rooted) ({ if (!list_push_back_inner(temproot, (Object*)_rooted)) { break ; } })
#define NEXT_NUM_UNSIGNED() ({ uint64_t _lit; if (!next_num_unsigned(closure->bytecode, &pointer, &_lit)) { error = exc_msg(&g_RuntimeError, "out of bounds"); break; } _lit; })
#define NEXT_NUM_SIGNED() ({ int64_t _lit; if (!next_num_signed(closure->bytecode, &pointer, &_lit)) { error = exc_msg(&g_RuntimeError, "out of bounds"); break; } _lit; })
#define NEXT_FLOAT() ({ double _lit; if (!next_float(closure->bytecode, &pointer, &_lit)) { error = exc_msg(&g_RuntimeError, "out of bounds"); break; } _lit; })
#define NEXT_OFFSET() ({ size_t _lit; if (!next_offset(closure->bytecode, &pointer, &_lit)) { error = exc_msg(&g_RuntimeError, "out of bounds"); break; } _lit; })
#define NEXT_BYTES() ({ BytesUnownedObject *_lit = next_bytes(closure->bytecode, &pointer); if (_lit == NULL) { error = exc_msg(&g_RuntimeError, "out of bounds"); break; } _lit; })
#define TEMP_ARGS0() ({ TupleObject *_tmp = tuple_raw(NULL, 0); if (_tmp == NULL) { break; } TEMPROOT(_tmp); _tmp; })
#define TEMP_ARGS1(arg1) ({ TupleObject *_tmp = tuple_raw((Object*[]){arg1}, 1); if (_tmp == NULL) { break; } TEMPROOT(_tmp); _tmp; })
#define TEMP_ARGS2(arg1, arg2) ({ TupleObject *_tmp = tuple_raw((Object*[]){arg1, arg2}, 2); if (_tmp == NULL) { break; } TEMPROOT(_tmp); _tmp; })
#define TEMP_ARGS3(arg1, arg2, arg3) ({ TupleObject *_tmp = tuple_raw((Object*[]){arg1, arg2, arg3}, 3); if (_tmp == NULL) { break; } TEMPROOT(_tmp); _tmp; })
#define BINOP(methodname) { Object *arg2 = POP(); Object *arg1 = POP(); Object *method = CHECK(get_attr_inner(arg1, "__" #methodname "__")); PUSH(CHECK(call(method, TEMP_ARGS1(arg2)))); }
#define UNOP(methodname) { Object *arg1 = POP(); Object *method = CHECK(get_attr_inner(arg1, "__" #methodname "__")); PUSH(CHECK(call(method, TEMP_ARGS0()))); }

			case ST_SWAP: {
				Object *a1 = POP();
				Object *a2 = POP();
				PUSH(a1);
				PUSH(a2);
				continue;
			}
			case ST_POP: {
				POP();
				continue;
			}
			case ST_DUP: {
				Object *a1 = POP();
				PUSH(a1);
				PUSH(a1);
				continue;
			}
			case ST_DUP2: {
				Object *a2 = POP();
				Object *a1 = POP();
				PUSH(a1);
				PUSH(a2);
				PUSH(a1);
				PUSH(a2);
				continue;
			}
			case LIT_BYTES: {
				PUSH(NEXT_BYTES());
				continue;
			}
			case LIT_INT: {
				PUSH(int_raw(NEXT_NUM_SIGNED()));
				continue;
			}
			case LIT_FLOAT: {
				PUSH(float_raw(NEXT_FLOAT()));
				continue;
			}
			case LIT_SLICE: {
				Object *end = POP();
				Object *start = POP();
				PUSH(slice_raw(start, end));
				continue;
			}
			case LIT_NONE: {
				PUSH(&g_none);
				continue;
			}
			case LIT_TRUE: {
				PUSH(&g_true);
				continue;
			}
			case LIT_FALSE: {
				PUSH(&g_false);
				continue;
			}
			case TUPLE_0:
			case TUPLE_1:
			case TUPLE_2:
			case TUPLE_3:
			case TUPLE_4:
			case TUPLE_N: {
				size_t count;
				if (opcode == TUPLE_N) {
					count = NEXT_NUM_UNSIGNED();
				} else {
					count = opcode - TUPLE_0;
				}
				if (stack->len < count) {
					error = exc_msg(&g_RuntimeError, "stack underflow");
					break;
				}
				TupleObject *result = tuple_raw(&stack->data[stack->len - count], count);
				if (result == NULL) {
					break;
				}
				for (size_t i = 0; i < count; i++) {
					POP();
				}
				PUSH(result);
				continue;
			}
			case CLOSURE: {
				Object *code = POP();
				if (code->type != &g_bytes) {
					error = exc_msg(&g_TypeError, "Expected bytes");
					break;
				}
				PUSH(CHECK(closure_raw((BytesObject*)code, locals)));
				continue;
			}
			case CLOSURE_BIND: {
				Object *code = POP();
				uint64_t num_idents = NEXT_NUM_UNSIGNED();
				if (code->type != &g_bytes) {
					error = exc_msg(&g_TypeError, "Expected bytes");
					break;
				}

				DictObject *new_context = CHECK(dicto_raw());
				TEMPROOT(new_context);
				for (uint64_t i = 0; i < num_idents; i++) {
					BytesUnownedObject *name = NEXT_BYTES();
					Object *value = CHECK(dict_getitem(TEMP_ARGS2((Object*)locals, (Object*)name)));
					CHECK(dict_setitem(TEMP_ARGS3((Object*)new_context, (Object*)name, value)));
				}

				PUSH(CHECK(closure_raw((BytesObject*)code, new_context)));
				continue;
			}
			case EMPTY_DICT: {
				PUSH(CHECK(dicto_raw()));
				continue;
			}
			case CLASS: {
				Object *dict = POP();
				Object *base = POP();
				PUSH(CHECK(type_constructor((Object*)&g_type, TEMP_ARGS2(base, dict))));
				continue;
			}
			case GET_ATTR: {
				Object *name = POP();
				Object *obj = POP();
				PUSH(CHECK(get_attr(obj, name)));
				continue;
			}
			case SET_ATTR: {
				Object *value = POP();
				Object *name = POP();
				Object *obj = POP();
				CHECK(set_attr(obj, name, value));
				continue;
			}
			case DEL_ATTR: {
				Object *name = POP();
				Object *obj = POP();
				CHECK(!del_attr(obj, name));
				continue;
			}
			case GET_ITEM: {
				Object *key = POP();
				Object *obj = POP();
				TEMPROOT(obj);
				Object *getattr = CHECK(get_attr_inner(obj, "__getitem__"));
				PUSH(CHECK(call(getattr, TEMP_ARGS1(key))));
				continue;
			}
			case SET_ITEM: {
				Object *val = POP();
				Object *key = POP();
				Object *obj = POP();
				TEMPROOT(obj);
				Object *setattr = CHECK(get_attr_inner(obj, "__setitem__"));
				CHECK(call(setattr, TEMP_ARGS2(key, val)));
				continue;
			}
			case DEL_ITEM: {
				Object *key = POP();
				Object *obj = POP();
				TEMPROOT(obj);
				Object *delattr = CHECK(get_attr_inner(obj, "__delitem__"));
				CHECK(call(delattr, TEMP_ARGS1(key)));
				continue;
			}
			case GET_LOCAL: {
				Object *name = POP();
				PUSH(CHECK(dict_getitem(TEMP_ARGS2((Object*)locals, name))));
				continue;
			}
			case SET_LOCAL: {
				Object *val = POP();
				Object *name = POP();
				CHECK(dict_setitem(TEMP_ARGS3((Object*)locals, name, val)));
				continue;
			}
			case DEL_LOCAL: {
				Object *name = POP();
				CHECK(dict_popitem(TEMP_ARGS2((Object*)locals, name)));
				continue;
			}
			case LOAD_ARGS: {
				PUSH(args);
				continue;
			}
			case JUMP: {
				pointer = &bytes_data(closure->bytecode)[NEXT_OFFSET()];
				continue;
			}
			case JUMP_IF: {
				size_t off = NEXT_OFFSET();
				Object *cond = POP();
				TEMPROOT(cond);
				Object *boolfunc = CHECK(get_attr_inner(cond, "__bool__"));
				Object *evaluated_cond = CHECK(call(boolfunc, TEMP_ARGS0()));
				if (evaluated_cond->type != &g_bool) {
					error = exc_msg(&g_TypeError, "__bool__ did not return bool");
					break;
				}
				if (evaluated_cond == (Object*)&g_true) {
					pointer = &bytes_data(closure->bytecode)[off];
				}
				continue;
			}
			case TRY: {
				size_t catch_target = NEXT_OFFSET();
				CHECK(list_push_back_inner(trystack, (Object*)int_raw((int64_t)catch_target)));
				continue;
			}
			case TRY_END: {
				CHECK(list_pop_back_inner(trystack));
				continue;
			}
			case CALL: {
				Object *args = POP();
				Object *target = POP();
				if (args->type != &g_tuple) {
					error = exc_msg(&g_TypeError, "Expected tuple");
					break;
				}
				TEMPROOT(args);
				TEMPROOT(target);
				PUSH(CHECK(call(target, (TupleObject*)args)));
				continue;
			}
			case SPAWN: {
				Object *args = POP();
				Object *target = POP();
				if (args->type != &g_tuple) {
					error = exc_msg(&g_TypeError, "Expected tuple");
					break;
				}
				TEMPROOT(args);
				TEMPROOT(target);
				PUSH(CHECK(thread_raw(target, (TupleObject*)args, &g_thread)));
				continue;
			}
			case RAISE: {
				error = POP();
				break;
			}
			case RETURN: {
				result = POP();
				goto EXIT;
			}
			case YIELD: {
				Object *val = POP();
				CHECK(thread_yield(val));
				continue;
			}
			case RAISE_IF_NOT_STOP: {
				Object *e = POP();
				if (!isinstance_inner(e, &g_StopIteration)) {
					error = e;
					break;
				}
				continue;
			}
			case OP_ADD: {
				BINOP(add);
				continue;
			}
			case OP_SUB: {
				BINOP(sub);
				continue;
			}
			case OP_MUL: {
				BINOP(mul);
				continue;
			}
			case OP_DIV: {
				BINOP(div);
				continue;
			}
			case OP_MOD: {
				BINOP(mod);
				continue;
			}
			case OP_AND: {
				BINOP(and);
				continue;
			}
			case OP_OR: {
				BINOP(or);
				continue;
			}
			case OP_XOR: {
				BINOP(xor);
				continue;
			}
			case OP_NEG: {
				UNOP(neg);
				continue;
			}
			case OP_NOT: {
				UNOP(not);
				continue;
			}
			case OP_INV: {
				UNOP(inv);
				continue;
			}
			case OP_EQ: {
				BINOP(eq);
				continue;
			}
			case OP_NE: {
				BINOP(ne);
				continue;
			}
			case OP_GT: {
				BINOP(gt);
				continue;
			}
			case OP_LT: {
				BINOP(lt);
				continue;
			}
			case OP_GE: {
				BINOP(ge);
				continue;
			}
			case OP_LE: {
				BINOP(le);
				continue;
			}
			case OP_SHL: {
				BINOP(shl);
				continue;
			}
			case OP_SHR: {
				BINOP(shr);
				continue;
			}
			default: {
				error = exc_msg(&g_RuntimeError, "Bad opcode");
				break;
			}
		}

		Object *local_error, *_catch_target;
ERROR:
		local_error = error;
		_catch_target = list_pop_back_inner(trystack);
		if (_catch_target == NULL) {
			error = local_error;
			goto EXIT;
		}
		if (_catch_target->type != (TypeObject*)&g_int) {
			puts("Fatal error: catch target was not int");
			exit(1);
		}
		size_t catch_target = ((IntObject*)_catch_target)->value;
		pointer = &bytes_data(closure->bytecode)[catch_target];

		while (stack->len) {
			POP(); // ok to use this outside the switch
		}
		if (!list_push_back_inner(stack, error)) {
			error = (Object*)&MemoryError_inst;
			// disard the entire try stack and just bail out
			goto EXIT;
		}
		continue;
	}

EXIT:
	gc_unroot((Object*)stack);
	gc_unroot((Object*)locals);
	gc_unroot((Object*)temproot);
	gc_unroot((Object*)trystack);
	return result;
}

File: src/interpreter.rs
use core::mem;
use std::{
    convert::TryInto,
    io::{Cursor, Read},
};

use byteorder::ReadBytesExt;
use leb128;
use num_enum::TryFromPrimitive;

use crate::{
    gcell::GCellOwner,
    object::{
        BytesObject, DictObject, ExceptionObject, Object, Result, TupleObject, G_BYTES, GcGCellExt,
        MEMORYERROR_INST, IntObject, FloatObject, G_NONE, G_TRUE, G_FALSE,
    },
};

enum StepResult {
    Normal,
    TryPush(i32),
    TryPop,
}

fn vec_reserve<T>(v: &mut Vec<T>, n: usize) -> Result<()> {
    v.try_reserve(n).map_err(|_| MEMORYERROR_INST.clone())
}

fn fallible_push<T>(vec: &mut Vec<T>, val: T) -> Result<()> {
    vec_reserve(vec, 1)?;
    vec.push(val);
    Ok(())
}

fn fallible_pop<T>(vec: &mut Vec<T>) -> Result<T> {
    if let Some(r) = vec.pop() {
        Ok(r)
    } else {
        return Err(ExceptionObject::runtime_error("Stack underflow")?);
    }
}

#[allow(non_camel_case_types)]
#[derive(TryFromPrimitive)]
#[repr(u8)]
enum Opcode {
    ERROR = 0,
    ST_SWAP = 1,
    ST_POP = 2,
    ST_DUP = 3,
    ST_DUP2 = 4,
    LIT_BYTES = 10,
    LIT_INT = 11,
    LIT_FLOAT = 12,
    LIT_SLICE = 13,
    LIT_NONE = 14,
    LIT_TRUE = 15,
    LIT_FALSE = 16,
    TUPLE_0 = 17,
    TUPLE_1 = 18,
    TUPLE_2 = 19,
    TUPLE_3 = 20,
    TUPLE_4 = 21,
    TUPLE_N = 22,
    CLOSURE = 23,
    CLOSURE_BIND = 24,
    EMPTY_DICT = 25,
    CLASS = 26,
    GET_ATTR = 40,
    SET_ATTR = 41,
    DEL_ATTR = 42,
    GET_ITEM = 43,
    SET_ITEM = 44,
    DEL_ITEM = 45,
    GET_LOCAL = 46,
    SET_LOCAL = 47,
    DEL_LOCAL = 48,
    LOAD_ARGS = 49,
    JUMP = 60,
    JUMP_IF = 61,
    TRY = 62,
    TRY_END = 63,
    CALL = 64,
    SPAWN = 65,
    RAISE = 66,
    RETURN = 67,
    YIELD = 68,
    RAISE_IF_NOT_STOP = 69,
    OP_ADD = 80,
    OP_SUB = 81,
    OP_MUL = 82,
    OP_DIV = 83,
    OP_MOD = 84,
    OP_AND = 85,
    OP_OR = 86,
    OP_XOR = 87,
    OP_NEG = 88,
    OP_NOT = 89,
    OP_INV = 90,
    OP_EQ = 91,
    OP_NE = 92,
    OP_GT = 93,
    OP_LT = 94,
    OP_GE = 95,
    OP_LE = 96,
    OP_SHL = 97,
    OP_SHR = 98,
}

fn interpreter(code: &BytesObject, context: &DictObject, args: &TupleObject, gil: &GCellOwner) {
    let mut locals = context.clone();
    let mut pc = 0;
    loop {}
}

fn interpreter_inner(
    gil: &mut GCellOwner,
    code: &mut Cursor<&[u8]>,
    locals: &mut DictObject,
    stack: &mut Vec<Object>,
) -> Result<StepResult> {
    let op = next_opcode(code)?;
    match op {
        Opcode::ERROR => {
            Err(ExceptionObject::runtime_error("Invalid opcode")?)
        }
        Opcode::ST_SWAP => {
            if let [.., first, second] = stack.as_mut_slice() {
                mem::swap(first, second);
                Ok(StepResult::Normal)
            } else {
                Err(ExceptionObject::runtime_error("Stack underflow")?)
            }
        }
        Opcode::ST_POP => {
            fallible_pop(stack)?;
            Ok(StepResult::Normal)
        }
        Opcode::ST_DUP => {
            if let Some(obj) = stack.last().cloned() {
                fallible_push(stack, obj)?;
                Ok(StepResult::Normal)
            } else {
                Err(ExceptionObject::runtime_error("Stack underflow")?)
            }
        }
        Opcode::ST_DUP2 => {
            if let Some(st2) = stack.get(stack.len().wrapping_sub(2)).cloned() {
                fallible_push(stack, st2)?;
                Ok(StepResult::Normal)
            } else {
                Err(ExceptionObject::runtime_error("Stack underflow")?)
            }
        }
        Opcode::LIT_BYTES => {
            let val = next_bytes(code)?;
            fallible_push(stack, val.into())?;
            Ok(StepResult::Normal)
        }
        Opcode::LIT_INT => {
            let val = IntObject::new(next_signed(code)?)?;
            fallible_push(stack, val.into())?;
            Ok(StepResult::Normal)
        }
        Opcode::LIT_FLOAT => {
            let val = FloatObject::new(next_float(code)?)?;
            fallible_push(stack, val.into())?;
            Ok(StepResult::Normal)
        }
        Opcode::LIT_SLICE => {
            let val2 = fallible_pop(stack)?;
            let val1 = fallible_pop(stack)?;
            todo!()
        }
        Opcode::LIT_NONE => {
            fallible_push(stack, G_NONE.clone().into())?;
            Ok(StepResult::Normal)
        }
        Opcode::LIT_TRUE => {
            fallible_push(stack, G_TRUE.clone().into())?;
            Ok(StepResult::Normal)
        }
        Opcode::LIT_FALSE => {
            fallible_push(stack, G_FALSE.clone().into())?;
            Ok(StepResult::Normal)
        }
        Opcode::TUPLE_0 => {
            fallible_push(stack, TupleObject::new0()?.into())?;
            Ok(StepResult::Normal)
        }
        Opcode::TUPLE_1 => {
            let val = fallible_pop(stack)?;
            fallible_push(stack, TupleObject::new1(val)?.into())?;
            Ok(StepResult::Normal)
        }
        Opcode::TUPLE_2 => {
            let val2 = fallible_pop(stack)?;
            let val1 = fallible_pop(stack)?;
            fallible_push(stack, TupleObject::new2(val1, val2)?.into())?;
            Ok(StepResult::Normal)
        }
        Opcode::TUPLE_3 => {
            let val3 = fallible_pop(stack)?;
            let val2 = fallible_pop(stack)?;
            let val1 = fallible_pop(stack)?;
            fallible_push(stack, TupleObject::new3(val1, val2, val3)?.into())?;
            Ok(StepResult::Normal)
        }
        Opcode::TUPLE_4 => {
            let val4 = fallible_pop(stack)?;
            let val3 = fallible_pop(stack)?;
            let val2 = fallible_pop(stack)?;
            let val1 = fallible_pop(stack)?;
            fallible_push(stack, TupleObject::new4(val1, val2, val3, val4)?.into())?;
            Ok(StepResult::Normal)
        }
        Opcode::TUPLE_N => {
            let len = next_unsigned(code)? as usize;
            if len > stack.len() {
                Err(ExceptionObject::runtime_error("Stack underflow")?)
            } else {
                let mut vals = vec![];
                vec_reserve(&mut vals, len)?;
                vals.extend_from_slice(&stack[stack.len() - len..]);
                stack.truncate(stack.len() - len);
                fallible_push(stack, TupleObject::new(vals)?.into())?;
                Ok(StepResult::Normal)
            }
        }
        Opcode::CLOSURE => {
            todo!()
        }
        Opcode::CLOSURE_BIND => {
            todo!()
        }
        Opcode::EMPTY_DICT => {
            todo!()
        }
        Opcode::CLASS => {
            todo!()
        }
        Opcode::GET_ATTR => {
            let name_obj = fallible_pop(stack)?;
            let obj = fallible_pop(stack)?;
            let result = obj.get_attr(gil, name_obj.as_ref())?;
            fallible_push(stack, result)?;
            Ok(StepResult::Normal)
        }
        Opcode::SET_ATTR => {
            let value = fallible_pop(stack)?;
            let name_obj = fallible_pop(stack)?;
            let obj = fallible_pop(stack)?;
            obj.set_attr(gil, name_obj.as_ref(), value)?;
            Ok(StepResult::Normal)
        }
        Opcode::DEL_ATTR => {
            todo!()
        }
        Opcode::GET_ITEM => {
            todo!()
        }
        Opcode::SET_ITEM => {
            todo!()
        }
        Opcode::DEL_ITEM => {
            todo!()
        }
        Opcode::GET_LOCAL => {
            todo!()
        }
        Opcode::SET_LOCAL => {
            todo!()
        }
        Opcode::DEL_LOCAL => {
            todo!()
        }
        Opcode::LOAD_ARGS => {
            todo!()
        }
        Opcode::JUMP => {
            todo!()
        }
        Opcode::JUMP_IF => {
            todo!()
        }
        Opcode::TRY => {
            todo!()
        }
        Opcode::TRY_END => {
            todo!()
        }
        Opcode::CALL => {
            todo!()
        }
        Opcode::SPAWN => {
            todo!()
        }
        Opcode::RAISE => {
            todo!()
        }
        Opcode::RETURN => {
            todo!()
        }
        Opcode::YIELD => {
            todo!()
        }
        Opcode::RAISE_IF_NOT_STOP => {
            todo!()
        }
        Opcode::OP_ADD => {
            todo!()
        }
        Opcode::OP_SUB => {
            todo!()
        }
        Opcode::OP_MUL => {
            todo!()
        }
        Opcode::OP_DIV => {
            todo!()
        }
        Opcode::OP_MOD => {
            todo!()
        }
        Opcode::OP_AND => {
            todo!()
        }
        Opcode::OP_OR => {
            todo!()
        }
        Opcode::OP_XOR => {
            todo!()
        }
        Opcode::OP_NEG => {
            todo!()
        }
        Opcode::OP_NOT => {
            todo!()
        }
        Opcode::OP_INV => {
            todo!()
        }
        Opcode::OP_EQ => {
            todo!()
        }
        Opcode::OP_NE => {
            todo!()
        }
        Opcode::OP_GT => {
            todo!()
        }
        Opcode::OP_LT => {
            todo!()
        }
        Opcode::OP_GE => {
            todo!()
        }
        Opcode::OP_LE => {
            todo!()
        }
        Opcode::OP_SHL => {
            todo!()
        }
        Opcode::OP_SHR => {
            todo!()
        }
    }
}

fn next_opcode(code: &mut Cursor<&[u8]>) -> Result<Opcode> {
    if code.is_empty() {
        Err(ExceptionObject::runtime_error("End of bytecode")?)
    } else {
        code.read_u8()
            .unwrap()
            .try_into()
            .or_else(|_| Err(ExceptionObject::runtime_error("Invalid opcode")?))
    }
}

fn next_float(code: &mut Cursor<&[u8]>) -> Result<f64> {
    if let Ok(f) = code.read_f64::<byteorder::LE>() {
        Ok(f)
    } else {
        Err(ExceptionObject::runtime_error("End of bytecode")?)
    }
}

fn next_unsigned(code: &mut Cursor<&[u8]>) -> Result<u64> {
    match leb128::read::unsigned(code) {
        Ok(len) => Ok(len),
        Err(leb128::read::Error::Overflow) => Err(ExceptionObject::overflow_error(
            "Literal integer too large",
        )?),
        Err(leb128::read::Error::IoError(_)) => {
            Err(ExceptionObject::runtime_error("End of bytecode")?)
        }
    }
}

fn next_signed(code: &mut Cursor<&[u8]>) -> Result<i64> {
    match leb128::read::signed(code) {
        Ok(len) => Ok(len),
        Err(leb128::read::Error::Overflow) => Err(ExceptionObject::overflow_error(
            "Literal integer too large",
        )?),
        Err(leb128::read::Error::IoError(_)) => {
            Err(ExceptionObject::runtime_error("End of bytecode")?)
        }
    }
}

fn next_bytes(code: &mut Cursor<&[u8]>) -> Result<BytesObject> {
    let len = next_unsigned(code)?;
    let mut buf = vec![0; len as usize];
    match code.read_exact(&mut buf) {
        Ok(()) => Ok(BytesObject {
            ty: G_BYTES.clone(),
            value: buf,
        }
        .into()),
        Err(_) => Err(ExceptionObject::runtime_error("End of bytecode")?),
    }
}
